File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode2438RangeProductQueriesOfPowers.cpp
#include "LeetCode2438RangeProductQueriesOfPowers.h"

#include "Auxilaries.h"
using namespace std;
int moduloPow(int power,int modulo)
{
    long result=1;
    int currentPower=1;
    long current{2};
    while (currentPower<=power)
    {
        if (currentPower&power)
        {
            result=(result*current)%modulo;
        }
        current*=current;
        current%=modulo;
        currentPower <<=1;
    }
    return static_cast<int>(result);
}
vector<int> LeetCode2438RangeProductQueriesOfPowers::productQueries(int n, vector<vector<int>>& queries)
{
    const long modulo=1e9+7;
    vector<int> powOfTwos{};
    long powOfTwo{1};
    int power{};
    while (powOfTwo<=n)
    {
        if ((n&powOfTwo)==powOfTwo)
        {
            powOfTwos.push_back(power);
        }
        if (powOfTwo>=n)
            break;
        powOfTwo <<=1;
        power++;
    }
    int powTwoLength=static_cast<int>(powOfTwos.size());
    for (int i=1;i<powTwoLength;i++)
    {
        powOfTwos[i]+=powOfTwos[i-1];
    }
    int length=static_cast<int>(queries.size());
    vector<int> result(length,0);
    for (int i=0;i<length;i++)
    {
        int lowerBound=queries[i][0]>0?powOfTwos[ queries[i][0]-1] : 0;
        int pow= powOfTwos[ queries[i][1]]-lowerBound;
        result[i]=moduloPow(pow,modulo);
    }
    return result;
}

void LeetCode2438RangeProductQueriesOfPowers::Run()
{
    vector<vector<int>> queries{{0,1},{2,2},{0,3}};
    int n{15};
    Auxilaries::PrintVector(productQueries(n,queries));
}

File: LeetCodeProblemsGo/LeetCode2438RangeProductQueriesOfPowers/LeetCode2438RangeProductQueriesOfPowers.go
package LeetCode2438RangeProductQueriesOfPowers

func moduloPow(power int, modulo int64) int {
	var current int64 = 1
	var currentPow int64 = 2
	pow := 1
	for pow <= power {
		if (pow & power) == pow {
			current = (currentPow * current) % modulo
			current %= modulo
		}
		pow <<= 1
		currentPow = (currentPow * currentPow) % modulo
	}
	return int(current)
}
func productQueries(n int, queries [][]int) []int {
	currentPow := 1
	pow := 0
	twoPows := []int{0}
	for currentPow <= n {
		if (n & currentPow) == currentPow {
			twoPows = append(twoPows, pow)
		}
		currentPow <<= 1
		pow++
	}
	var modulo int64 = 1e9 + 7
	for i := 1; i < len(twoPows); i++ {
		twoPows[i] += twoPows[i-1]
	}
	length := len(queries)
	result := make([]int, length)
	for i, query := range queries {
		delta := twoPows[query[1]+1] - twoPows[query[0]]
		result[i] = moduloPow(delta, modulo)
	}
	return result
}
